### PR TITLE
refactor(gateway): switch to undici for service-to-service communication

### LINKED ISSUE #263: Switch to undici for federation
https://github.com/nodejs/undici is significantly faster than nodejs http client.

### PR DESCRIPTION
**Pull request by PacoDu**
Fixes #263

### CODE REVIEW

File: lib/gateway/request.js [hunk 1]
```diff
@@ -1,54 +1,40 @@
...
 const eos = require('end-of-stream')
 const { FederatedError } = require('../errors')
 
 function agentOption (opts) {
   return {
     keepAlive: true,
-    keepAliveMsecs: opts.keepAliveMsecs || 60 * 1000, // 1 minute
-    maxSockets: opts.maxSockets || 2048,
-    maxFreeSockets: opts.maxFreeSockets || 2048,
-    rejectUnauthorized: opts.rejectUnauthorized
```
Comment (mcollina) [inline lib/gateway/request.js:13]: `rejectUnauthorized` should be passed in `tls.rejectUnauthorized`.

File: lib/gateway/request.js [hunk 2]
```diff
@@ -1,54 +1,40 @@
...
-const https = require('https')
+const { Pool } = require('undici')
+const { URL } = require('url')
 const eos = require('end-of-stream')
 const { FederatedError } = require('../errors')
 
 function agentOption (opts) {
   return {
     keepAlive: true,
-    keepAliveMsecs: opts.keepAliveMsecs || 60 * 1000, // 1 minute
```
Comment (mcollina) [inline lib/gateway/request.js:10]: This should become `keepAliveMaxTimeout`

File: lib/gateway/request.js [hunk 3]
```diff
@@ -1,54 +1,40 @@
...
-    maxSockets: opts.maxSockets || 2048,
-    maxFreeSockets: opts.maxFreeSockets || 2048,
-    rejectUnauthorized: opts.rejectUnauthorized
+    keepAliveMaxTimeout: opts.keepAliveMsecs || 60 * 1000, // 1 minute
+    connections: opts.connections || 10
```
Comment (mcollina) [inline lib/gateway/request.js:11]:
```suggestion
    connections: opts.connections || opts.maxSockets || 10
```

Backward compatibility

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
semver-major or semver-minor?

**Comment by PacoDu** [comment]
> semver-major or semver-minor?

`keepAliveMsecs`and `maxFreeSockets` are now removed but they were undocumented, I'm not sure if this is a breaking change...

File: lib/gateway/request.js [hunk 4]
```diff
@@ -1,54 +1,43 @@
...
 const { FederatedError } = require('../errors')
 
 function agentOption (opts) {
   return {
     keepAlive: true,
-    keepAliveMsecs: opts.keepAliveMsecs || 60 * 1000, // 1 minute
-    maxSockets: opts.maxSockets || 2048,
-    maxFreeSockets: opts.maxFreeSockets || 2048,
-    rejectUnauthorized: opts.rejectUnauthorized
+    keepAliveMaxTimeout: opts.keepAliveMaxTimeout || 60 * 1000, // 1 minute
```
Comment (mcollina) [inline lib/gateway/request.js:10]:
```suggestion
    keepAliveMaxTimeout: opts.keepAliveMaxTimeout || opts.keepAliveMsecs || 60 * 1000, // 1 minute
```

**Comment by simoneb** [comment]
Are there any further changes needed to this PR or can it land? Also note that there's now a new major of undici.

**Comment by mcollina** [comment]
This needs an update to undici v3 :). I'm waiting to ship a new major here.

**Comment by PacoDu** [comment]
> This needs an update to undici v3 :). I'm waiting to ship a new major here.

Sure i'll take a look as soon as possible.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**